### PR TITLE
Fix backprop of `LinkAsTorchModel`

### PR DESCRIPTION
**Pull request by niboshi**
Continuation after #6

Fixes `LinkAsTorchModel` so that it can backprop through the underlying Chainer model.

### CODE REVIEW

**Comment by niboshi** [comment]
There seems to be a problem in grad manipulation. Working on a fix.

**Comment by niboshi** [comment]
During backprop in a Chainer model, some parameter's `.grad.array` is replaced (not in place) for some reason.
This causes desynchronization between Chainer grad and the corresponding Torch grad tensor that wraps it.

**Comment by niboshi** [comment]
Added `cpm.Optimizer` to handle grad synchronization.
PTAL

File: chainer_pytorch_migration/parameter.py [hunk 1]
```diff
@@ -16,12 +30,94 @@ class LinkAsTorchModel(nn.Module):
...
+    '''
+    Torch tensor from which backprop can be performed.
+    '''
+    def __new__(cls, variable):
+        assert isinstance(variable, chainer.Variable)
+        obj = cpm.astensor(variable.array)
+        obj.__class__ = cls
+        return obj
 
+    def __init__(self, variable):
```
Comment (emcastillo) [inline chainer_pytorch_migration/parameter.py]: I think `__init__` is not executed if `__new__` is defined, am I wrong?

File: chainer_pytorch_migration/parameter.py [hunk 2]
```diff
@@ -16,12 +30,94 @@ class LinkAsTorchModel(nn.Module):
...
+    def __getattr__(self, name):
+        if name in ('step', 'zero_grad', '_base_optimizer'):
+            return object.__getattribute__(self, name)
+        return getattr(self._base_optimizer, name)
+
+    def step(self, closure=None):
+        for param_group in self._base_optimizer.param_groups:
+            for param in param_group['params']:
+                assert isinstance(param, ChainerParameter)
+                param.grad.copy_(torch.tensor(param._param.grad))
```
Comment (emcastillo) [inline chainer_pytorch_migration/parameter.py]: could the grad temporal copy be avoided by doing
`param.grad.copy_(cpm.astensor(param._param.grad))`?

Comment (niboshi) [reply chainer_pytorch_migration/parameter.py]: 👍 Thanks, fixed in b968d1767a714d2e206806d67fc87306096f5209

**Pull request merged**